### PR TITLE
[sp] : fix the attention kernel for sp

### PR DESCRIPTION
**Pull request by wangbluo**
## 📝 What does this PR do?

For cases where s_q * s_kv * element size >= 10 GB, dispatch only to the FlashAttentionDaoLoader kernel, and use an empty tensor as a placeholder for the attention_mask. Additionally, only causal and padded causal modes are supported.

### CODE REVIEW

File: colossalai/shardformer/layer/attn.py [hunk 1]
```diff
@@ -163,7 +172,7 @@ def prepare_attn_kwargs(
             outputs["attention_mask_type"] = AttnMaskType.CAUSAL
             attention_mask = torch.ones(s_q, s_kv, dtype=dtype, device=device)
             if s_q != 1:
-                attention_mask = attention_mask.tril(diagonal=0)
+                attention_mask.tril_(diagonal=0)
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:175]: Don't generate attention mask here if mask exceeds memory bound

Comment (ver217) [reply colossalai/shardformer/layer/attn.py:175]: Move line 210-217 here

Comment (ver217) [reply colossalai/shardformer/layer/attn.py:175]: And move to "else" branch.

File: colossalai/shardformer/layer/attn.py [hunk 2]
```diff
@@ -278,8 +296,16 @@ def attention(
             assert attention_mask_type == AttnMaskType.CUSTOM
 
         # kernel dispatch
+        b, _, s_q, _ = q.shape
+        b, _, s_kv, _ = v.shape
+        element_size = torch.tensor([], dtype=q.dtype).element_size()
+        memory_size = s_q * s_kv * element_size
+        if memory_size > MEMORY_BOUND:
+            attention_mask = torch.empty((0,), dtype=q.dtype, device=q.device)
+            assert attention_mask_type == AttnMaskType.PADDED_CAUSAL or AttnMaskType.PADDED
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:305]: Don't overwrite attention mask here. We only need memory size here.

File: colossalai/shardformer/layer/attn.py [hunk 3]
```diff
@@ -113,12 +116,18 @@ def _dispatch_kernel(dtype: torch.dtype, mask_type: Optional[AttnMaskType]) -> C
...
+        if size > MEMORY_BOUND:
+            FlashAttentionDaoLoader().load()
         # lazy load
         if isinstance(ColoAttention._kernel_dispatch_map[dtype][mask_type], KernelLoader):
             ColoAttention._kernel_dispatch_map[dtype][mask_type] = ColoAttention._kernel_dispatch_map[dtype][
                 mask_type
             ].load()
-        return ColoAttention._kernel_dispatch_map[dtype][mask_type]
+
+        return (
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:128]: `if size >= MEMORY_BOUND and mask_type in (AttnMaskType.PADDED_CAUSAL, AttnMaskType.CAUSAL)`

File: colossalai/shardformer/layer/attn.py [hunk 4]
```diff
@@ -197,6 +206,15 @@ def prepare_attn_kwargs(
...
             attention_mask = invert_mask(attention_mask).unsqueeze(1)
         outputs["attention_mask"] = attention_mask
+
+        element_size = torch.tensor([], dtype=dtype).element_size()
+        memory_size = s_q * s_kv * element_size
+        if memory_size > MEMORY_BOUND:
+            attention_mask = torch.empty((0,), dtype=dtype, device=device)
+            outputs["attention_mask"] = attention_mask
+            if outputs["attention_mask_type"] != AttnMaskType.PADDED_CAUSAL:
+                outputs["attention_mask_type"] = AttnMaskType.CAUSAL
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:216]: If memory exceeds bound and the mask type is padded/custom, then let it oom. Becuase dao's flash-attn does not support these two modes.

File: colossalai/shardformer/layer/attn.py [hunk 5]
```diff
@@ -278,8 +299,15 @@ def attention(
             assert attention_mask_type == AttnMaskType.CUSTOM
 
         # kernel dispatch
+        b, _, s_q, _ = q.shape
+        b, _, s_kv, _ = v.shape
+        element_size = torch.tensor([], dtype=q.dtype).element_size()
+        memory_size = s_q * s_kv * element_size
+        if memory_size > MEMORY_BOUND:
+            assert attention_mask_type == AttnMaskType.PADDED_CAUSAL or AttnMaskType.PADDED
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:307]: remove assert

File: colossalai/shardformer/layer/attn.py [hunk 6]
```diff
@@ -177,7 +192,10 @@ def prepare_attn_kwargs(
                 b,
                 s_kv,
             ), f"Padding mask shape {kv_padding_mask.shape} should align with shape 4d ({b}, {s_kv})"
-            attention_mask = kv_padding_mask[:, None, :].expand(b, s_q, s_kv).to(dtype=dtype, device=device)
+            if memory_size > MEMORY_BOUND:
+                attention_mask = torch.empty((0,), dtype=dtype, device=device)
+            else:
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:197]: If memory <= MEMORY_BOUND or not is_causal, then expand...

File: colossalai/shardformer/layer/attn.py [hunk 7]
```diff
@@ -113,12 +116,19 @@ def _dispatch_kernel(dtype: torch.dtype, mask_type: Optional[AttnMaskType]) -> C
             raise ValueError(
                 "FlashAttention kernel is not available for dtype {} and mask_type {}".format(dtype, mask_type)
             )
+
+        if size > MEMORY_BOUND:
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:120]: Use the same condition. Line 128 uses `>=`, but got `>` here.

**Review by ver217 — APPROVED** [review]

File: colossalai/shardformer/layer/attn.py [hunk 8]
```diff
@@ -113,12 +116,19 @@ def _dispatch_kernel(dtype: torch.dtype, mask_type: Optional[AttnMaskType]) -> C
             raise ValueError(
                 "FlashAttention kernel is not available for dtype {} and mask_type {}".format(dtype, mask_type)
             )
+
+        if size >= MEMORY_BOUND:
+            flash_kernel = FlashAttentionDaoLoader().load()
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:121]: save this to class attribute
